include/Oscillator.h

```diff
@@ -153,6 +164,63 @@ class LMMS_EXPORT Oscillator
 		return m_userWave->userWaveSample( _sample );
 	}
 
+	struct wtSampleControl {
+		float frame;
+		f_cnt_t f1;
+		f_cnt_t f2;
+		int band;
+	};
+
+	inline wtSampleControl getWtSampleControl(const float sample)
+	{
+		wtSampleControl control;
+		control.frame = sample * OscillatorConstants::WAVETABLE_LENGTH;
+		control.f1 = static_cast<f_cnt_t>(control.frame) % OscillatorConstants::WAVETABLE_LENGTH;
+		if (control.f1 < 0)
+		{
+			control.f1 += OscillatorConstants::WAVETABLE_LENGTH;
+		}
+		control.f2 = control.f1 < OscillatorConstants::WAVETABLE_LENGTH - 1 ?
+					control.f1 + 1 :
+					0;
+		control.band = waveTableBandFromFreq(m_freq * m_detuning * Engine::mixer()->processingSampleRate());
+		return control;
+	}
+
+	inline sample_t wtSample(const sample_t table[OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT][OscillatorConstants::WAVETABLE_LENGTH], const float _sample)
```

Comment:
I'm also not quite clear on why is it needed, but when I remove it and add a `const sample_t **table` version, the compiler still complains that:
`no known conversion for argument 1 from ‘sample_t [128][2446]’ {aka ‘float [128][2446]’} to ‘sample_t**’ {aka ‘float**’}`

And here I thought that pointers and arrays are always supposed to be interchangeable..

Reply:
I think [it's not possible](https://stackoverflow.com/questions/8203700/conversion-of-2d-array-to-pointer-to-pointer)? Looks like you need to give at most one dimension to the type, because otherwise, it does not know how to increment a pointer to `table`. Does it work if you specify only one dimension (and do the other by array-brackets)?

Reply:
Huh, right, that makes sense. If it's just a pointer to pointer, there are no meta-data describing the dimensions. It seems that:
`const sample_t table[][OscillatorConstants::WAVETABLE_LENGTH]`
is working, so I will use that. The other way around it does not work, which makes sense, considering the way how arrays are stored in memory (described in the second stackoverflow answer).

Reply:
I wonder if your first version is even better (with both sizes specified). This one allows the compiler to make boundary checks for the first version. From what I learned from Dom in discord now, all these array things are decayed to pointers.

Reply:
So basically return to the original version and remove the newly added assert? Since I presume that if the "arrayness" is preserved, apart from allowing range checks, it should also ensure it won't be a `nullptr`, right? The only exception I can think of would be if someone intentionally casts `nullptr` as the 2D array, which is... why would anyone do that?

Reply:
I just wrote `void func(s_t s[2][3]) { assert(s); }` and called `func(nullptr);`. No compiler warnings, and the assert terminated the program. Same with passing `ptr` after declaring it as `s_t (*ptr)[3] = nullptr;`. I would prefer to keep the asserts, though I guess not having them will not kill us :D

Reply:
All right then, I will return the long version (while trying to add some line break in there) and keep all asserts. The three overloads doing all essentially the same thing still seem a bit weird and redundant, but dealing with this little piece of code is taking way too much time already, so I would rather just add the asserts and call it finished. ^^;

Reply:
agreed :+1: